### PR TITLE
Ticket 46941: Data entry grid view height recalculate

### PR DESCRIPTION
**Pull request by labkey-martyp**
#### Rationale
Grid views in data entry do not always correctly re-set height as the contents change or are updated. This update just ensures the grid view height is set to the height of it's body to ensure all contents are displayed. Covers initial load, adding/removing rows and multi-line text cell edits

#### Changes
* Add resize height function
* Add updateLayout override to adjust height of grid view to contents
* Add resize after validation complete

### CODE REVIEW

**Review by labkey-jeckels — APPROVED** [review]

**Comment by bbimber** [comment]
Hey - 

I believe this bug exists, but have two questions: 

1) do you have a sense on what kind of overhead all this re-rendering causes? Does resizing the height trigger a full redo of the layout, meaning this might introduce a lot more repeat doLayout calls? If true, that's potentially a non-trivial change. If you're trying the grid with a handful of rows, this probably isnt even noticeable, and would work fine. However, when working on grids/forms with a lot of rows (perhaps daily observations / rounds, and/or some other workflow where entry involves whole rooms), then this might matter a lot more. I dont know what bookkeeping is associated with changing the view height, for example. If there is non-trivial work, is there a low-cost way to check the current height and compare to the new value to limit when you call this?

2) In general I appreciate trying to limit the scope of the change (i.e. EHR vs. LDK). Do you have a sense on whether the problem of rows getting out-of-sync with height is EHR-specific (perhaps due to the background validation), or is it a more general issue Ext4 grids can have? My sense is that it was caused by text-wrapping, and perhaps due to Ext4's layout/height calc not handling this optimally. But it was a long time ago when I really looked at this.

**Comment by labkey-martyp** [comment]
> Hey -
> 
> I believe this bug exists, but have two questions:
> 
> 1. do you have a sense on what kind of overhead all this re-rendering causes? Does resizing the height trigger a full redo of the layout, meaning this might introduce a lot more repeat doLayout calls? If true, that's potentially a non-trivial change. If you're trying the grid with a handful of rows, this probably isnt even noticeable, and would work fine. However, when working on grids/forms with a lot of rows (perhaps daily observations / rounds, and/or some other workflow where entry involves whole rooms), then this might matter a lot more. I dont know what bookkeeping is associated with changing the view height, for example. If there is non-trivial work, is there a low-cost way to check the current height and compare to the new value to limit when you call this?
> 2. In general I appreciate trying to limit the scope of the change (i.e. EHR vs. LDK). Do you have a sense on whether the problem of rows getting out-of-sync with height is EHR-specific (perhaps due to the background validation), or is it a more general issue Ext4 grids can have? My sense is that it was caused by text-wrapping, and perhaps due to Ext4's layout/height calc not handling this optimally. But it was a long time ago when I really looked at this.

1. Yeah initially I tried only doing this when view and view body heights didn't match but there are some cases like bulk update that are editing the grid directly where the heights match on the components but not in the display. Calling setHeight does the bookkeeping to trigger the layout update. That being said, as far as I can tell this doesn't increase the number of actual layout updates on the component. The setHeight does only that then refers to the parent to call updateLayout which loops back through this component to call the normal updateLayout. At least as far as AbstractComponent.updateLayout calls they are the same. But it's a good point to try this out on a larger set of records. I'll do that before merging.
2. This is definitely due to some manipulation of data after the grid render and possibly during validation. I didn't dig too deep into trying to fix an individual form, instead doing the more generic approach to make the grid view more reactive to the contents. I'm not sure it's EHR specific other than there are a lot of varying requirements and customizations.

**Comment by bbimber** [comment]
OK, thanks @labkey-martyp. After xmas I might take a look into this for non-EHR grids. The ones I know about a a lot less demanding than EHR, but they might encounter this too. If that turns out to be the case, I might suggest pushing this into the base LDK class.

File: ehr/resources/web/ehr/grid/Panel.js
Comment (bbimber) [inline ehr/resources/web/ehr/grid/Panel.js]: If this method has arguments, should this be: this.callParent(arguments)?

**Review by bbimber — CHANGES_REQUESTED** [review]
That's great you were able to simplify this into a relatively discrete change in the grid component alone. Minor comment above.

File: ehr/resources/web/ehr/grid/Panel.js
Comment (bbimber) [reply ehr/resources/web/ehr/grid/Panel.js]: actually - so this only reaches callParent() when heightResize=false? So if heightResize=true, then we never call the parent class? Is that a good thing?

Comment (labkey-martyp) [reply ehr/resources/web/ehr/grid/Panel.js]: Good call on the arguments. Updated that.

Ultimately setHeight calls setSize which calls updateLayout on the component again, which is ultimately why we need the heightResize flag to avoid an infinite loop.  https://docs.sencha.com/extjs/4.2.1/source/AbstractComponent.html#Ext-AbstractComponent-method-setSize

**Comment by labkey-martyp** [comment]
@bbimber Let me know if you have any more thoughts on this Ben. Otherwise it needs your approval to get merged in.

**Comment by bbimber** [comment]
@labkey-martyp I wont have a any to do any real investigation until next week. If you want to merge, i just approved. I am going to do a little looking to see if i can repro this and if it makes sense in LDK, but I dont have any specific comments beyond what I already added. I dont understand the infinite loop issue enough without digging into code myself.

**Review by bbimber — APPROVED** [review]

**Pull request merged**
